analyzers/src/SonarAnalyzer.Common/AnalysisContext/SonarReportingContextBase.cs

```diff
@@ -28,7 +28,7 @@ public abstract class SonarReportingContextBase<TContext> : SonarAnalysisContext
 
     protected void ReportIssueCore(Diagnostic diagnostic)
     {
-        if (HasMatchingScope(diagnostic.Descriptor))
+        if (HasMatchingScope(diagnostic.Descriptor) && SonarAnalysisContext.LegacyIsRegisteredActionEnabled(new[] { diagnostic.Descriptor }, diagnostic.Location?.SourceTree))
```

Comment:
The original (wrong) design was actually hiding this problem - that the conditions from registration need to be all re-checked at reporting time too.

The `ShouldAnalyzeTree` condition is checked in the derived classes below.

Reply:
`ReportIssueCore` is called quite often. You may add a small optimization to avoid the creation of the array:
```suggestion
        if (HasMatchingScope(diagnostic.Descriptor) && SonarAnalysisContext.ShouldExecuteRegisteredAction != null && SonarAnalysisContext.LegacyIsRegisteredActionEnabled(new[] { diagnostic.Descriptor }, diagnostic.Location?.SourceTree))
```

Reply:
> that the conditions from registration need to be all re-checked at reporting time too.

Is this for the case that we end up in another SyntaxTree and report there (e.g. via `ISymbol.DeclaringSyntaxReferences`)? Do we have tests for this? If not, can you add one?

Reply:
It is for a case when we register for a symbol action, or a compilation action.
During registration, no `SyntaxTree` is available. So action is registered and executed.
During issue reporting, we (most likely) have the reliable `SourceTree` of the issue itself. And we should take the decision based on that one.

Also, in theory, we can have two rules implemented in the same analyzer. If rule 1 is suppressed by the legacy behavior and rule 2 is not, the registration (having rule 1 & rule 2) should return true, while reporting (having rules separately) should provide precise answer.

Reply:
There was an UT that failed without this. So it is tested.

Reply:
I've duplicated the legacy method to push the array creation down. And to keep the usages of `ShouldExecuteRegisteredAction` isolated on one place